src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,540 @@
+/// items will fail to hydrate.
+/// TODO(jkosh44) Back-fill these tables in read-only mode so they can properly hydrate.
+async fn migrate_builtin_items_0dt(
+    state: &mut CatalogState,
+    txn: &mut Transaction<'_>,
+    persist_client: PersistClient,
+    migrated_builtins: Vec<GlobalId>,
+    deploy_generation: u64,
+    read_only: bool,
+) -> Result<BuiltinItemMigrationResult, Error> {
+    // 1. Open migration shard.
+    let organization_id = state.config.environment_id.organization_id();
+    let shard_id = builtin_migration_shard_id(organization_id);
+    let diagnostics = Diagnostics {
+        shard_name: "builtin_migration".to_string(),
+        handle_purpose: format!("builtin table migration shard for org {organization_id:?} generation {deploy_generation:?}"),
+    };
+    let mut since_handle: SinceHandle<TableKey, ShardId, Timestamp, Diff, i64> = persist_client
+        .open_critical_since(
+            shard_id,
+            // TODO: We may need to use a different critical reader
+            // id for this if we want to be able to introspect it via SQL.
+            PersistClient::CONTROLLER_CRITICAL_SINCE,
+            diagnostics.clone(),
+        )
+        .await
+        .expect("invalid usage");
+    let (mut write_handle, mut read_handle): (
+        WriteHandle<TableKey, ShardId, Timestamp, Diff>,
+        ReadHandle<TableKey, ShardId, Timestamp, Diff>,
+    ) = persist_client
+        .open(
+            shard_id,
+            Arc::new(TableKeySchema),
+            Arc::new(ShardIdSchema),
+            diagnostics,
+            USE_CRITICAL_SINCE_CATALOG.get(persist_client.dyncfgs()),
+        )
+        .await
+        .expect("invalid usage");
+    // Commit an empty write at the minimum timestamp so the shard is always readable.
+    const EMPTY_UPDATES: &[((TableKey, ShardId), Timestamp, Diff)] = &[];
+    let _ = write_handle
+        .compare_and_append(
+            EMPTY_UPDATES,
+            Antichain::from_elem(Timestamp::minimum()),
+            Antichain::from_elem(Timestamp::minimum().step_forward()),
+        )
+        .await
+        .expect("invalid usage");
+
+    // 2. Get the `GlobalId` of all migrated tables.
+    let migrated_tables: BTreeSet<_> = migrated_builtins
+        .into_iter()
+        .filter_map(|id| {
+            if state.get_entry(&id).item().is_storage_collection() {
+                match id {
+                    GlobalId::System(id) => Some(id),
+                    _ => unreachable!("builtin objects must have system ID, found: {id:?}"),
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // 3. Read in the current contents of the migration shard.
+    let upper = fetch_upper(&mut write_handle).await;
+    let as_of = Antichain::from_elem(upper.saturating_sub(1));
+    let since = read_handle.since();
+    as_of.join(since);
```

Comment:
Dunno how much we care if this is semi-permanent code, but needing to consider both the upper and the since when we compute the as_of feels a touch sloppy to me. Better would be a protocol where we're sure we only have to look at one of them.

I'm fine if you ignore this :)

Reply:
We only ever downgrade the since after writing to the shard, and we set the since to `upper - 1`. So it should be enough to remove the line `as_of.join(since);` and replace it with something like `assert!(since <= as_of);`. Does that check out?

Reply:
Yup! We'd also want to note that we intentionally open the read handle before we fetch the upper, to address races between the upper and since moving forward in some other process

Reply:
Done